### PR TITLE
RFC: Adds Technical planning document

### PR DESCRIPTION
**Pull request by sweir27**
# Proposal
This RFC proposes:
- Adding a new document describing our technical planning process. This will be linked to from the existing `technology_choices` document, but since tech plans are about more than just adding new technologies, we feel it warrants its own page.
- Clarification to all steps of our current tech planning process, much of which is implicit. We expect these steps to change over time. This clarification touches topics such as:
    - What is a tech plan
    - When to write one
    - How to write one
    - What the possible states are (and how to advance between them)
    - How to choose stakeholders
    - How to provide feedback and approve a tech plan
- Consolidating a few fields on the template:
    - `Contributors` and `Champions` will be removed, replaced by a single field “Author(s)” so that it’s clear who to contact about a given tech plan. We expect this to be filled by 1-2 people.
    - Removing the `DRI` and `Decider` fields. We found these fields are used differently by different team members and add confusion. We instead propose an updated “Stakeholders” section which should cover this need.
- Adding a new state: Implemented, to clarify the difference between a tech plan that was accepted (but then deprioritized) vs. one whose changes exist in code.
- Expanding the Stakeholders section to include different review states: Approved, Rejected, and Changes Requested (with a comment field) in order to clarify the review process.

# Reasoning
As our team has grown and more tech plans have been written, it’s become clear that some clarification of our existing process would help streamline the process. The lack of clarity was confirmed in a recent survey we conducted with the engineering team. While most respondents appreciated the open nature of our tech planning process and how widely it has been adopted, there were a lot of questions, particularly around the states and approval process.

The goal of this RFC is to make our tech planning process clearer. By writing down what our existing process is, it should also be easier to update via similar RFCs to this one.

# How is this RFC resolved?
If there is a strong majority of 👍 compared to 👎 and at least 20 votes, this RFC will be resolved.

There are two parts to this RFC resolution:
- Merging in this PR (taking comments into account), which adds the new document and links to it from a few places in README.
- Updating the tech plan template based on the proposal above.

### CODE REVIEW

File: playbooks/technical-planning.md
Comment (damassi) [inline playbooks/technical-planning.md:26]: Should we clarify when to start a tech plan? Should we wait until there's a formal (and signed-off) product brief? Or (as in the paragraph below) do they tend to work symbiotically? To have a fully spec'd-out product brief seems ideal, but at times it doesn't seem entirely practical if the target seems to be moving. 

Is there a situation where a technical plan is written _without_ a product brief? And if so, when?

File: playbooks/technical-planning.md
Comment (damassi) [inline playbooks/technical-planning.md:46]: It feels slightly confusing to have multiple state spread over different sections of the tech plan. Why not aggregate `Approved`, `Rejected`, and `Changes Requested` here vs under stakeholders? Is it because one stakeholder might have requested changes while another might approve?

Comment (damassi) [reply playbooks/technical-planning.md:46]: If anything, it might be worthwhile to have a single additional state added here -- `Changes Requested` denoting that a stakeholder has requested changes. This puts the tech plan in a high-level holding pattern.

File: playbooks/technical-planning.md
Comment (damassi) [inline playbooks/technical-planning.md:71]: 👍

File: playbooks/technical-planning.md
Comment (joeyAghion) [inline playbooks/technical-planning.md:19]: 👌

File: playbooks/technical-planning.md
Comment (joeyAghion) [inline playbooks/technical-planning.md:68]:
```suggestion
Significant tech plans benefit from a synchronous review in a practice meeting. If there is not a practice meeting that works for your timeline, schedule something ad-hoc. Make sure to mention the tech plan in [our weekly engineering standup](https://github.com/artsy/README/blob/main/events/open-standup.md) to invite broader feedback.
```

File: playbooks/technical-planning.md
Comment (joeyAghion) [inline playbooks/technical-planning.md:61]: Sometimes I volunteer feedback despite not being invited as a stakeholder. This could even be critical/blocking feedback. Should I add myself as a stakeholder in order to capture that status formally? (Or would that be off-putting to authors?)

**Review by joeyAghion — APPROVED** [review]

File: playbooks/technical-planning.md
Comment (cavvia) [reply playbooks/technical-planning.md:61]: I think this _could_ continue to be a source of tension—no provision is made in the process for the possibility that someone outside of your stakeholder group may have a good reason to request blocking changes. A rotating board is designed to even out such influence over tech plans, as we would also like to avoid individuals having special privileges in the process.

File: playbooks/technical-planning.md
Comment (leamotta) [inline playbooks/technical-planning.md:40]: I think I would find useful an additional state -- `Implementing`, specially for scenarios where a technical plan involves architectural or eng-wide changes. It would give us a better overview into what the tech plan status is, as `Accepted` does not mean that there is active work going on

File: playbooks/technical-planning.md
Comment (sweir27) [reply playbooks/technical-planning.md:61]: My first thought is that if you have blocking feedback, especially if significant, you probably _should_ have been added a as a stakeholder in the first place. So it doesn't bother me for us to suggest you add yourself as one? But curious what others think. 

Also intentionally aligning this with how we do PRs – if you have blocking feedback on a PR, do you usually add a review? Or just leave comments indicating the feedback is blocking?

File: playbooks/technical-planning.md
Comment (iskounen) [reply playbooks/technical-planning.md:26]: 1. In my experience tech plans are easiest to write and review when the desired outcome is clear. Sometimes the desired outcome of a project brief is clear before it receives sign-off, and sometimes it's not clear even after it receives sign-off. I would avoid creating a dependency between technical plans and project briefs based on the state of a project brief for this reason. I think that the following paragraphs do a good job of explaining how engineers should react in a case where the desired outcome is unclear.
2. I think that tech plans should originate from RFCs, practice meetings, future fridays, etc. as well as from product briefs.

File: playbooks/technical-planning.md
Comment (iskounen) [inline playbooks/technical-planning.md:64]: **question:** What do you mean by "We recommend filling out the Feedback requested by field."?

File: playbooks/README.md
Comment (iskounen) [inline playbooks/README.md:51]: **nitpick:** I think the "at Artsy" is redundant.

File: playbooks/technical-planning.md
Comment (lean8086) [inline playbooks/technical-planning.md:8]: In my case, writing tech plans have also been extremely useful to understand how things currently work. Those things are usually spread across old PRs, slack threads, or in people's minds.

File: playbooks/technical-planning.md
Comment (lean8086) [inline playbooks/technical-planning.md:28]: These (positively) happened many times to me. I think it's one of the main advantages of working on tech plans. I would consider mentioning it as part of the list "Writing tech plans helps us"!

File: playbooks/technical-planning.md
Comment (lean8086) [inline playbooks/technical-planning.md:43]: It would be nice to add to the closed tech plan a little description of why it's been closed, so we also have a reference to why the decision has been made.

File: playbooks/technical-planning.md
Comment (nickskalkin) [inline playbooks/technical-planning.md:23]: 👍 I would stress that it is always better to document a decision (even a small one), especially if it might be not obvious for people who will work on this in future. Short [Y-statements](https://medium.com/olzzio/y-statements-10eb07b5a177) can be enough for this purpose.

File: playbooks/technical-planning.md
Comment (damassi) [reply playbooks/technical-planning.md:43]: Indeed, similar to a full description on a revert PR. Very helpful 👍

File: playbooks/technical-planning.md
Comment (lean8086) [reply playbooks/technical-planning.md:34]: I saw [Adam's comment](https://github.com/artsy/README/pull/483/files#r1021420395) on changing Proposed to Open for Feedback, and that would clarify things.

File: playbooks/technical-planning.md
Comment (joeyAghion) [reply playbooks/technical-planning.md:23]: Hadn't heard of Y-statements or ([architecture haikus](https://www.georgefairbanks.com/blog/comparch-wicsa-2011-panel-discussion-and-haiku-tutorial/) 😄) before. Cool!

File: playbooks/technical-planning.md
Comment (dblandin) [inline playbooks/technical-planning.md:39]: **suggestion**: Would we consider technical plan states that aren't so interconnected with stakeholder review states?

Inspired by the interaction between GitHub Pull Requests and Review Requests, we could have something like this:

## Technical Plans

```mermaid
flowchart LR
  Draft <-- ready for feedback --> Open
  Draft --> Closed
  Open -- acceptance criteria met --> Implementing
  Open --> Closed
  Implementing -- executed --> Done
  Implementing ---> Closed
  style Draft fill:#FFF689
  style Open fill:#7CA5B8
  style Closed fill:#EC0B43
  style Implementing fill:#FFA400
  style Done fill:#84B082
```

## Review Requests

```mermaid
flowchart LR
  Pending <--> Approved
  Pending <--> ChangesRequested(Changes Requested)
  style Pending fill:#7CA5B8
  style ChangesRequested fill:#EC0B43
  style Approved fill:#84B082
```

File: playbooks/technical-planning.md
Comment (dblandin) [inline playbooks/technical-planning.md:78]: **question**: Wondering if this rejected state could be folded within "Changes Requested". If the proposal is completely unacceptable, the change requested could be to rethink the entire approach — at which point the technical plan is either closed or put back into a draft state.

File: playbooks/technical-planning.md
Comment (dblandin) [inline playbooks/technical-planning.md:60]: **question**: If we're shifting this list to usually include engineers, would it make sense to rename this to "Reviewers" rather than "Stakeholders"?

File: playbooks/technical-planning.md
Comment (dblandin) [reply playbooks/technical-planning.md:34]: **suggestion**: +1 to clarifying these states as they relate to feedback. In my experience, I treat Draft as a state where the plan is initially forming and "feedback only upon request". With Proposed, an invitation for broader (and unsolicited) feedback is welcome.

File: playbooks/technical-planning.md
Comment (chr-tatu) [reply playbooks/technical-planning.md:64]: Is it about a date aka Feedback requested by <date>?

File: playbooks/technical-planning.md
Comment (chr-tatu) [inline playbooks/technical-planning.md:68]: I find it better to announce tech plans on the #dev channel, too. I noticed some folks do it already, but not everyone. For instance, some folks may be OoO during the engineering stand-up. Having a slack message will allow everyone to see what they missed while they were away.

File: playbooks/technical-planning.md
Comment (mc-jones) [inline playbooks/technical-planning.md:21]: Are we able to define a bit more what we mean by "best practices" here? Putting myself in the shoes of a well-meaning but new Artsy engineer, I am not sure I would be able to make this evaluation without more guidance on what is considered best practice.

File: playbooks/technical-planning.md
Comment (mc-jones) [reply playbooks/technical-planning.md:23]: Maybe a bit off-topic, but I've always liked our own [engineering playbook's approach](https://github.com/artsy/README/blob/main/playbooks/engineer-workflow.md#pull-requests) to this, which encourages a similar (but lightweight version) of this for PRs.

File: playbooks/technical-planning.md
Comment (mc-jones) [inline playbooks/technical-planning.md:81]: What happens if a stakeholder (or `reviewer`, depending on what we end up calling them) marks something as `Rejected`? Who gets the final say as to if the plan is implemented or not in this scenario? (Sorry if I just missed this somewhere!).

File: playbooks/technical-planning.md
Comment (mc-jones) [reply playbooks/technical-planning.md:61]: Similar to my [comment below](https://github.com/artsy/README/pull/483/files#r1030801473), do we need to clarify how we will resolve/escalate disagreements? While my hope is that most disagreements can be handled between the author(s) and the reviewer making blocking feedback, if we do allow reviewers to add themselves, I can envision a scenario where someone adds themselves as a reviewer and makes blocking feedback causing conflict. I think having a clear process for handling situations like this will help avoid tech plans which solicit strong opinions from becoming blocked.

File: playbooks/technical-planning.md
Comment (joeyAghion) [reply playbooks/technical-planning.md:21]: I'd say "established patterns" instead, which are at least reliable and safe within a given project.

File: playbooks/technical-planning.md
Comment (sweir27) [reply playbooks/technical-planning.md:64]: Ah yes – right now there's a "Feedback requested by" field (I can update to include the quotes).

File: playbooks/technical-planning.md
Comment (sweir27) [reply playbooks/technical-planning.md:81]: Good question – based on the feedback above, I think I'll remove the "Rejected" status in favor of "Changes requested". I'd say the writer is still responsible for the ultimate decision, but they are expected to take the feedback from reviewers seriously and addressing the requested changes.

File: playbooks/technical-planning.md
Comment (sweir27) [reply playbooks/technical-planning.md:55]: Could this be folded _into_ the stakeholders section? Like, ensure that at least one member of the affected team is added to that list (with a little explanation for why?). Totally open to adding a new section, just wondering how it would behave differently.

File: playbooks/technical-planning.md
Comment (sweir27) [reply playbooks/technical-planning.md:23]: I can leave a note here about the importance of documenting decisions (at least somewhere!).

File: playbooks/technical-planning.md
Comment (sweir27) [reply playbooks/technical-planning.md:26]: +1 to what @iskounen said!

File: playbooks/technical-planning.md
Comment (sweir27) [reply playbooks/technical-planning.md:61]: @mc-jones You bring up a good point. I'm hesitant to add more process in this iteration vs. dealing with (and learning from) conflicts as they arise. My hope is that by making the tech plans, feedback, and review statuses clearer, it'll be easier to discuss and understand different points of view, and will allow the tech plan writer to feel empowered to make calls (vs. delegating that responsibility elsewhere). I suspect most disagreements can be figured out with a face-to-face between those involved, and maybe we can rely on our standard approach to escalations (raising to your manager) if it becomes more sensitive. How does that sound?

File: playbooks/technical-planning.md
Comment (sweir27) [reply playbooks/technical-planning.md:39]: Ooh I like this. So we remove the "Approved" state in favor of "Implementing".

File: playbooks/technical-planning.md
Comment (mc-jones) [reply playbooks/technical-planning.md:61]: @sweir27 Sounds good!

**Review by mc-jones — APPROVED** [review]
🙌

**Comment by sweir27** [comment]
Thanks to all for the comments! I've addressed as many as I could. There are still a few outstanding discussions – I encourage people to follow up with subsequent PRs (or with follow-up discussion in #dev)!

After merging in, I'll update the tech plan template.

**Pull request merged**
